lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -2677,24 +2687,60 @@ void IRGenDebugInfoImpl::emitVariableDeclaration(
     LocalVarCache.insert({Key, llvm::TrackingMDNodeRef(Var)});
   }
 
-  auto appendDIExpression =
-      [&VarInfo, this](llvm::DIExpression *DIExpr) -> llvm::DIExpression * {
-    if (VarInfo.DIExpr) {
-      llvm::SmallVector<uint64_t, 2> Operands;
-      if (!buildDebugInfoExpression(VarInfo, Operands))
-        return nullptr;
-      if (Operands.size())
-        return llvm::DIExpression::append(DIExpr, Operands);
-    }
-    return DIExpr;
-  };
-
   // Running variables for the current/previous piece.
   bool IsPiece = Storage.size() > 1;
   uint64_t SizeOfByte = CI.getTargetInfo().getCharWidth();
   unsigned AlignInBits = SizeOfByte;
   unsigned OffsetInBits = 0;
   unsigned SizeInBits = 0;
+  std::pair<unsigned, unsigned> Fragment(0, 0);
+
+  auto appendDIExpression =
+      [&VarInfo, this](
+          llvm::DIExpression *DIExpr,
+          std::pair<unsigned, unsigned> PieceFragment) -> llvm::DIExpression * {
+    unsigned PieceFragmentOffset = PieceFragment.first;
+    unsigned PieceFragmentSize = PieceFragment.second;
+
+    if (!VarInfo.DIExpr) {
+      if (!PieceFragmentSize)
+        return DIExpr;
+
+      return llvm::DIExpression::createFragmentExpression(
+                 DIExpr, PieceFragmentOffset, PieceFragmentSize)
+          .getValueOr(nullptr);
+    }
+
+    llvm::SmallVector<uint64_t, 2> Operands;
+    std::pair<unsigned, unsigned> VarFragment(0, 0);
+    if (!buildDebugInfoExpression(VarInfo, Operands, VarFragment))
+      return nullptr;
+
+    if (!Operands.empty())
+      DIExpr = llvm::DIExpression::append(DIExpr, Operands);
+
+    unsigned VarFragmentOffset = VarFragment.first;
+    unsigned VarFragmentSize = VarFragment.second;
+
+    // Add the fragment of the SIL variable.
+    if (VarFragmentSize)
+      DIExpr = llvm::DIExpression::createFragmentExpression(
+                   DIExpr, VarFragmentOffset, VarFragmentSize)
+                   .getValueOr(nullptr);
+
+    if (!DIExpr)
+      return nullptr;
+
+    // When the fragment of the SIL variable is further split into other
+    // fragments (PieceFragment), merge them into one DW_OP_LLVM_Fragment
+    // expression.
+    if (PieceFragmentSize)
```

Comment:
Should we assert that PieceFragmentSize < VarFragmentSize here?

Reply:
This is checked by `createFragmentExpression`:
```
        assert((OffsetInBits + SizeInBits <= FragmentSizeInBits) &&
               "new fragment outside of original fragment");
```
where `FragmentSizeInBits` is the fragment expression we're appending to.